### PR TITLE
Fix compressed offset bug

### PR DESCRIPTION
**Pull request by klippx**
Addresses #505 

We are happy to change the test, since we were not sure how to set up the test scenario properly we used the "raw" approach of simply using data from our system console.

### CODE REVIEW

**Comment by dasch** [comment]
This should be updated to reflect https://github.com/zendesk/ruby-kafka/issues/505#issuecomment-352826114

File: lib/kafka/protocol/message.rb [hunk 1]
```diff
@@ -58,10 +58,14 @@ def decompress
         message_set_decoder = Decoder.from_string(data)
         message_set = MessageSet.decode(message_set_decoder)
 
+        max_relative_offset = message_set.messages.last.offset
+        return message_set if max_relative_offset == offset
+
         # The contained messages need to have their offset corrected.
+        base_offset = offset - max_relative_offset
         messages = message_set.messages.each_with_index.map do |message, i|
```
Comment (zmstone) [inline lib/kafka/protocol/message.rb]: i no longer need

File: lib/kafka/compressor.rb [hunk 2]
```diff
@@ -28,14 +28,15 @@ def initialize(codec_name: nil, threshold: 1, instrumenter:)
 
     # @param message_set [Protocol::MessageSet]
     # @return [Protocol::MessageSet]
-    def compress(message_set)
+    def compress(message_set, offset: -1)
```
Comment (dasch) [inline lib/kafka/compressor.rb]: Can you document that `offset` is just used to simulate the broker behavior?

File: lib/kafka/protocol/message.rb [hunk 3]
```diff
@@ -58,10 +58,14 @@ def decompress
         message_set_decoder = Decoder.from_string(data)
         message_set = MessageSet.decode(message_set_decoder)
 
+        max_relative_offset = message_set.messages.last.offset
+        return message_set if max_relative_offset == offset
```
Comment (dasch) [inline lib/kafka/protocol/message.rb]: This definitely needs a comment explaining why that's the case :)

File: spec/compressor_spec.rb [hunk 4]
```diff
@@ -2,29 +2,6 @@
...
-      decoded_message_set = decoded_message.decompress
-      messages = decoded_message_set.messages
-
-      expect(messages.map(&:value)).to eq ["hello1", "hello2"]
-
-      # When decoding a compressed message, the offsets are calculated relative to that
-      # of the container message. The broker will set the offset in normal operation,
-      # but at the client-side we set it to -1.
-      expect(messages.map(&:offset)).to eq [-1, 0]
-    end
```
Comment (dasch) [inline spec/compressor_spec.rb]: Why remove this?

**Comment by dasch** [comment]
Thanks for the PR! In order to ensure this doesn't break in the future, I'd love to get a few comments in the code describing the behavior.

File: spec/compressor_spec.rb (hunk 4 above)
Comment (klippx) [reply spec/compressor_spec.rb]: We moved tests to message_set_spec, and we beleived this case was already covered in the other three tests we provided. This test does has -1 as wrapper message offset, and both messages are -1, -1 as well. Hence our code will accept this as "correct from broker side" and return `[-1, -1]`

In the end we just thought the other test cases was more realistic. Should we add this back?

Comment (dasch) [reply spec/compressor_spec.rb]: OK – no, that's fine, just needed to follow along.

**Comment by dasch** [comment]
Looks good! Can you verify that this works with real workloads?

File: lib/kafka/protocol/message.rb [hunk 5]
```diff
@@ -113,6 +102,37 @@ def self.decode(decoder)
...
+      # Offsets may be relative with regards to wrapped message offset, but there are special cases.
+      #
+      # Cases when client will receive corrected offsets:
+      #   - When fetch request is version 0, kafka will correct relative offset on broker side before replying fetch response
+      #   - When messages is stored in 0.9 format on disk (broker configured to do so).
+      #
+      # All other cases, compressed inner messages should have relative offset, with below attributes:
+      #   - The container message should have the 'real' offset
+      #   - The container message's offset should be the 'real' offset of the last message in the compressed batch
+      #   - The first inner message should always have offset = 0
```
Comment (zmstone) [inline lib/kafka/protocol/message.rb]: sorry, I was wrong about this.
It doesn’t always start with 0

Comment (klippx) [reply lib/kafka/protocol/message.rb]: thanks, amended 👍🏼

**Comment by klippx** [comment]
@dasch We are going on christmas vacation right now, so we wont be able to work on this for about 3 weeks. So, this may take a while unfortunately i.e. creating a long feedback loop. Up to you. We are running 0.5.0 while we are on vacation which is stable for us. 

Besides, even if we did run this version our consumers would use the short-circuit of not fixing the offsets since last message offset == wrapped message offset so it wont be a fair test.

**Comment by dasch** [comment]
@klippx a change of this magnitude requires some real-life testing before I dare merge it – let's wait until you're back.

**Comment by zmstone** [comment]
This pr is about the same magnitude as 42821e9f
To be fair, that change was not covered by real-life testing either.
I'm not saying that tests are not important,
just that all consumers running current master code against compressed messages
are committing wrong offsets and likely skipping messages.
This change may have bug, but should not make things worse.

**Comment by piavka** [comment]
totally agree with @zmstone

**Comment by dasch** [comment]
OK, seems like I have a volunteer who can test it. I'll merge and cut a pre-release.

**Pull request merged**